sync-core/src/main/java/com/cloudant/sync/datastore/AttachmentManager.java

```diff
@@ -348,47 +382,153 @@ protected void purgeAttachments(SQLDatabase db) {
-                            }
-                        } catch (SecurityException e) {
-                            logger.log(Level.WARNING, "SecurityException when trying to delete file " +
-
-                                    "from BLOB store: " + f.getAbsolutePath(), e);
+            c.close();
+
+            // Now iterate through the attachments_key_filename, deleting items we need to
+            // (both db row and file on disk).
+            File attachments = new File(attachmentsDir);
+            c = db.rawQuery(SQL_ATTACHMENTS_SELECT_KEYS_FILENAMES, null);
+            while (c.moveToNext()) {
+                String keyForFile = c.getString(0);
+
+                if (!currentKeys.contains(keyForFile)) {
+                    File f = new File(attachments, c.getString(1));
+                    try {
+                        boolean deleted = f.delete();
+                        if (deleted) {
+                            db.delete(ATTACHMENTS_KEY_FILENAME, "key = ?", new String[]{keyForFile});
+                        } else {
+                            logger.warning("Could not delete file from BLOB store: " +
+                                    f.getAbsolutePath());
                         }
+                    } catch (SecurityException e) {
+                        logger.log(Level.WARNING, "SecurityException when trying to delete file " +
+                                "from BLOB store: " + f.getAbsolutePath(), e);
                     }
                 }
-        }
+            }
+            c.close();
+
         } catch (SQLException e) {
             logger.log(Level.SEVERE, "Problem in purgeAttachments, executing SQL to fetch all attachment keys ", e);
         }finally {
-           DatabaseUtils.closeCursorQuietly(c);
+            DatabaseUtils.closeCursorQuietly(c);
         }
     }
 
-    private String keyToString(byte[] key) {
+    private static String keyToString(byte[] key) {
         return new String(new Hex().encode(key));
     }
 
-    private File fileFromKey(byte[] key) {
-        return new File(attachmentsDir, keyToString(key));
+    /**
+     * Lookup or create a on disk File representation of blob, in {@code db} using {@code key}.
+     *
+     * Existing attachments will have an entry in db, so the method just looks this up
+     * and returns the File object for the path.
+     *
+     * For new attachments, the {@code key} doesn't already have an associated filename,
+     * one is generated and inserted into the database before returning a File object
+     * for blob associated with {@code key}.
+     *
+     * @param db database to use.
+     * @param key key to lookup filename for.
+     * @return File object for blob associated with {@code key}.
+     */
+    File fileFromKey(SQLDatabase db, byte[] key, boolean allowCreateName) throws AttachmentException {
+
+        String keyString = keyToString(key);
+        String filename = null;
+
+        db.beginTransaction();
```

Comment:
I'm not sure it's guaranteed to be the case; I don't think there's harm in the sub-transaction?